packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart

```diff
@@ -0,0 +1,824 @@
+  ///
+  /// The [env] property is used to locate the project files that are checked.
+  ///
+  /// The [templatesDir] parameter is optional. If null, the tool's default
+  /// templates directory will be used.
+  ///
+  /// When [exitOnFail] is set to true, the [handleResults] and [attemptToolExit]
+  /// methods will exit the tool when this validator detects a recommended
+  /// change. This defaults to true.
+  DeferredComponentsSetupValidator(this.env, {
+    this.exitOnFail = true,
+    String title,
+    Directory templatesDir,
+  }) : _outputDir = env.projectDir
+        .childDirectory('build')
+        .childDirectory(kDeferredComponentsTempDirectory),
+      _inputs = <File>[],
+      _outputs = <File>[],
+      _title = title ?? 'Deferred components setup verification',
+      _templatesDir = templatesDir,
+      _generatedFiles = <String>[],
+      _modifiedFiles = <String>[],
+      _invalidFiles = <String, String>{},
+      _diffLines = <String>[];
+
+  /// The build environment that should be used to find the input files to run
+  /// checks against.
+  ///
+  /// The checks in this class are meant to be used as part of a build process,
+  /// so an environment should be available.
+  final Environment env;
+
+  /// When true, failed checks and tasks will result in [attemptToolExit]
+  /// triggering [throwToolExit].
+  final bool exitOnFail;
+
+  /// The name of the golden file that tracks the latest loading units
+  /// generated.
+  @visibleForTesting
+  static const String kDeferredComponentsGoldenFileName = 'deferred_components_golden.yaml';
+  /// The directory in the build folder to generate missing/modified files into.
+  @visibleForTesting
+  static const String kDeferredComponentsTempDirectory = 'android_deferred_components_setup_files';
+
+  final String _title;
+  final Directory _templatesDir;
+  final Directory _outputDir;
+  // Files that were newly generated by this validator.
+  final List<String> _generatedFiles;
+  // Existing files that were modified by this validator.
+  final List<String> _modifiedFiles;
+  // Files that were invalid and unable to be checked. These files are input
+  // files that the validator tries to read rather than output files the
+  // validator generates. The key is the file name and the value is the message
+  // or reason it was invalid.
+  final Map<String, String> _invalidFiles;
+  // Output of the diff task.
+  // TODO(garyq): implement the diff task.
+  final List<String> _diffLines;
+  // Tracks the new and missing loading units.
+  Map<String, dynamic> _goldenComparisonResults;
+  // Input and output Files used to generate a Depfile.
+  final List<File> _inputs;
+  final List<File> _outputs;
+
+  /// Returns a [Depfile] that includes all input dependencies and non-temporary
+  /// output files.
+  ///
+  /// This should only be called once all desired checks and tasks have been
+  /// executed.
+  Depfile get depfile => Depfile(_inputs, _outputs);
```

Comment:
These methods can be made fairly standalone, by having each method individually return its inputs/outputs - then letting the caller be responsible for combining them.

Reply:
So I actually want the class to handle the check result internally. I don't really want the caller to be handling the results of the validation since there are specific behaviors that should happen based off of the results of each check, which is handled by `displayResults`. Letting the caller combine/handle the check results directly can miss a lot of nuance and result in incorrect recommendations to the developer.

These checks should not really be run independently of the overall validator class.